cdap-app-fabric/src/main/java/io/cdap/cdap/support/services/SupportBundleService.java

```diff
@@ -0,0 +1,194 @@
+    File basePath = new File(localDir, uuid);
+
+    SupportBundleStatus supportBundleStatus =
+      new SupportBundleStatus(uuid, System.currentTimeMillis(), supportBundleConfiguration,
+                              CollectionState.IN_PROGRESS);
+    try {
+      // Puts all the files under the uuid path
+      File baseDirectory = new File(localDir);
+      DirUtils.mkdirs(baseDirectory);
+      int fileCount = DirUtils.list(baseDirectory).size();
+
+      // We want to keep consistent number of bundle to provide to customer
+      int folderMaxNumber = cConf.getInt(Constants.SupportBundle.MAX_FOLDER_SIZE);
+      if (fileCount >= folderMaxNumber) {
+        File oldFilesDirectory = getOldestFolder(baseDirectory);
+        deleteOldFolders(oldFilesDirectory);
+      }
+      DirUtils.mkdirs(basePath);
+
+      SupportBundleStatus finishBundleStatus =
+        new SupportBundleStatus(supportBundleStatus, "", CollectionState.FINISHED, System.currentTimeMillis());
+      addToStatus(finishBundleStatus, basePath.getPath());
+    } catch (Exception e) {
+      SupportBundleStatus failedBundleStatus =
+        new SupportBundleStatus(supportBundleStatus, e.getMessage(), CollectionState.FAILED,
+                                System.currentTimeMillis());
+      addToStatus(failedBundleStatus, basePath.getPath());
+      throw new IllegalArgumentException("Failed to generate support bundle ", e);
+    }
+    return uuid;
+  }
+
+  /**
+   * Gets oldest folder from the root directory
+   */
+  private File getOldestFolder(File baseDirectory) {
+    List<File> uuidFiles = DirUtils.listFiles(baseDirectory).stream()
+      .filter(file -> !file.getName().startsWith(".") && !file.isHidden() && file.isDirectory())
+      .collect(Collectors.toList());
+    return Collections.min(uuidFiles, Comparator.<File, Boolean>comparing(f1 -> {
+      try {
+        return getSingleBundleJson(f1).getStatus() != CollectionState.FAILED;
+      } catch (Exception e) {
+        throw new RuntimeException("Failed to get file status ", e);
+      }
+    }).thenComparing(File::lastModified));
+  }
+
+  /**
+   * Deletes old folders after certain number of folders exist
+   */
+  private void deleteOldFolders(@Nullable File oldFilesDirectory) {
+    try {
+      DirUtils.deleteDirectoryContents(oldFilesDirectory);
+    } catch (IOException e) {
+      LOG.error("Failed to clean up directory {}", oldFilesDirectory, e);
+    }
+  }
+
+  @Override
+  public void close() {
+    this.executorService.shutdown();
+  }
+
+  /**
+   * Update status file
+   */
+  private void addToStatus(SupportBundleStatus supportBundleStatus, String basePath) throws IOException {
+    try (FileWriter statusFile = new FileWriter(new File(basePath, SupportBundleFileNames.STATUS_FILE_NAME))) {
+      GSON.toJson(supportBundleStatus, statusFile);
+      statusFile.flush();
+    } catch (IOException e) {
+      throw new IOException("Failed to update status file ", e);
```

Comment:
Why do you need to catch and rethrow IOException?

Reply:
So that we can surface the error to handler and return the corresponding error code and message

Reply:
Yep, but it's not Handler business. It done by `CommonNettyHttpServiceBuilder` automatically in `io.cdap.cdap.common.HttpExceptionHandler`

Reply:
You don't change the code here. It would make sense if you thrown some `HttpErrorStatusProvider` exception